packages/flutter/lib/src/material/icon.dart

```diff
@@ -65,13 +75,37 @@ class Icon extends StatelessComponent {
       category = parts[0];
       subtype = parts[1];
     }
-    // TODO(eseidel): This clearly isn't correct.  Not sure what would be.
-    // Should we use the ios images on ios?
-    String density = 'drawable-xxhdpi';
-    String colorSuffix = _getColorSuffix(context);
+    String density = new _MaterialDensityQualifier(context).resolve();
+    String colorSuffix = new _MaterialColorQualifier(context: context, color: color).resolve();
+    return '$category/drawable-$density/ic_${subtype}_${colorSuffix}_${iconSize}dp.png';
+  }
+}
+
+class Icon extends StatelessComponent {
+  Icon({
+    Key key,
+    this.size: IconSize.s24,
+    this.icon: '',
+    this.color,
+    this.colorFilter
+  }) : super(key: key) {
+    assert(size != null);
+    assert(icon != null);
+  }
+
+  final IconSize size;
+  final String icon;
+  final IconThemeColor color;
+  final ColorFilter colorFilter;
+
+  Widget build(BuildContext context) {
     int iconSize = _kIconSize[size];
     return new AssetImage(
-      name: '$category/$density/ic_${subtype}_${colorSuffix}_${iconSize}dp.png',
+      name: new _MaterialIconResolver(
```

Comment:
If you create a new one every frame, why aren't the fields in that class "final"? I'm unclear as to the lifetime or intended behaviour of this object.

I recommend adding comments before each of these classes explaining who is expect to own it, and what the lifetime is.